### PR TITLE
roachprod: support overriding the --join flag and per node init

### PR DESCRIPTION
**Pull request by adityamaru**
This change adds support to override the --join flag when running
`roachprod start`. Previously, the first node would join itself and
subsequent nodes would join the first node. Now, we respect the --join
flag if present, otherwise we default to the original behavior.

This change also adds support for running `roachprod init` on a per node
basis. The default behaviour of `roachprod init <cluster>` will not
change, but if a node (or range of nodes) is specified after the cluster
name then we will attempt to init through those nodes. If the nodes
belong to the same cluster, the command will fail with "cluster already
init'ed".

Both these changes are motivated by the need for spinning up multiple
clusters across roachprod VMs.  To start two clusters one now do any of:

`roachprod start local:1 --skip-init`
`roachprod start local:2 --skip-init --args="--join <address of node 2 itself>"`
`roachprod init local`
`roachprod init local:2`

or

`roachprod start local:1`
`roachprod start local:2 --args="--join <address of node 2 itself>"`
`roachprod init local:2`

### CODE REVIEW

**Comment by adityamaru** [comment]
I'm happy to bike shed on the name and approach, I just wanted to get a sense of what it entailed. I intend to follow this up with a `roachprod start-tenant` command and then eventually the roachtest we want 🙂

**Comment by adityamaru** [comment]
ahh I just realized that since more than one node can run the initialization logic we won't be able to use `Printf` inside the `c.Parallel()` closure to log any longer.

**Review by irfansharif — COMMENTED** [review]
If I'm understanding right, we basically want to run multiple crdb clusters on the same roachprod cluster? My first instinct to do that would be something like:
```
roachprod start <cluster-name>:<nodes> -a '--join <address-to-one-of-the-nodes>' --skip-init
roachprod init <cluster-name>:<one-of-the-nodes>
```

This probably won't work today, but I think what I'm getting at is that we don't need to introduce `--new-kv-server`, do we? 

Also, I have no idea what the future of roachtests + multi-tenant integration tests would look like. Doesn't look like this `--new-kv-server` flag is doing anything other than mucking about with the join flags (I assume there's a way to only start cockroach in "kv server" mode, which we're no doing). Before we even open _that_ can of worms, given that @tbg is thinking about revamping roachtests entirely for 21.1, perhaps you could make do with my suggestion above and massaging the impl of `-a` and `--skip-init` to do what you want. I'll let Tobi chime in on the rest.

**Comment by adityamaru** [comment]
> If I'm understanding right, we basically want to run multiple crdb clusters on the same roachprod cluster? My first instinct to do that would be something like:
> 
> ```
> roachprod start <cluster-name>:<nodes> -a '--join <address-to-one-of-the-nodes>' --skip-init
> roachprod init <cluster-name>:<one-of-the-nodes>
> ```
> 
> This probably won't work today, but I think what I'm getting at is that we don't need to introduce `--new-kv-server`, do we?


I like this idea, I looked into this a little bit and it would entail:
- support overriding the  `--join` flag if specified.
- teaching `roachprod init` to operate on a per-node basis if specified after the cluster name. 

One UX quirk that pops up is, to maintain current behavior, `roachprod init local` would initialize nodeIdx 0 (instead of running init on each of the nodes. On the other hand, `roachprod init local:2` would initialize the second VM, and `roachprod init local:2-4` would init each of the VM from 2 to 4. I guess this is fine, as long as it's documented in the help text?

**Review by irfansharif — COMMENTED** [review]
>  On the other hand, roachprod init local:2 would initialize the second VM, and roachprod init local:2-4 would init each of the VM from 2 to 4

`roachprod init`, which calls into `cockroach init`, is only supposed to be run against one node in the cluster, just once. If you were to `roachprod init <cluster>:2-4`, if the nodes knew about one another (through join flags), 2 of the three init attempts would fail. If the nodes didn't know about one another, that'd end up creating three different clusters. All of that sounds fine to me, they're inline with the semantics of `cockroach init`.

**Comment by adityamaru** [comment]
Default behaviour, when every VM is part of the same cluster.
Screen Shot 2021-03-23 at 12 20 17 PM

Creating two clusters, rooted at node 1 and 2. Node 3 joins node 2's cluster and node 4 joins node 1's cluster.
Screen Shot 2021-03-23 at 12 32 58 PM


Screen Shot 2021-03-23 at 12 33 31 PM

Screen Shot 2021-03-23 at 12 33 36 PM

**Review by irfansharif — COMMENTED** [review]
Iterating through all the VMs and `cockroach init`-ing them is unsound, and not how we typically want operators to do it. I get that it "works" if we go through node 1 first, and no-op if we get back a "already initialized" error, but it's still unsound. Consider what happens if n2 doesn't hear about n1 being initialized (despite being --join-ed to it) and is also `init`-ed (because we're effectively racing with roachprod). That'll end up creating two clusters, when we wanted one. 

Is it not possible to keep the earlier behaviour where `roachprod init <cluster>` only targets n1, and if specified, `roachprod init <cluster>:<nX>` does it through nX?

**Comment by tbg** [comment]
Can I ask a meta question - are you writing a roachtest as the primary way to test tenant-to-tenant replication? I am asking because I would ask for an in-process test (TestCluster) to exist before you try to do the same in a roachtest. A roachtest makes sense to test large datasets (i.e. things you couldn't reasonably run on a laptop). To test the basic functionality, an in-process test is advised. This is both because the cost of ownership of a roachtest is typically fairly high (all of the flakes are the owners') and because the debug turnaround is much slower. I'm happy to engage further if you need advice on how to accomplish the same in a unit test.

**Comment by adityamaru** [comment]
> Can I ask a meta question - are you writing a roachtest as the primary way to test tenant-to-tenant replication? I am asking because I would ask for an in-process test (TestCluster) to exist before you try to do the same in a roachtest.

Nope, we do have in-process tests for _most_ components in `streamingccl`. We also have at least one e2e test, but all these tests write and ingest trivial amounts of data. 
The motivation for writing a roachtest was to be able to run the replication logic over a longer period of time with a workload running on the tenant to stress the streaming, ingestion, and cutover. As it stands the functionality is not production-ready but this lays the groundwork for seeing how the system behaves in the presence of workloads that are not O(seconds). We definitely plan to have in-process tests side by side.

As a side note, @irfansharif did mention a complete revamp of roachprod/roachtest in 21.2, I'm curious to know if those efforts are motivated by making it less costly to write roachtests or to move more of those operations in-process thereby raising the bar for what should be a roachtest? Apologies if I'm being hand wavy but I couldn't find much on this topic.

**Comment by tbg** [comment]
> Nope, we do have in-process tests for most components in streamingccl. We also have at least one e2e test, but all these tests write and ingest trivial amounts of data.

👍 cool, thanks, sounds good to go ahead with a roachtest then.

> As a side note, @irfansharif did mention a complete revamp of roachprod/roachtest in 21.2, I'm curious to know if those efforts are motivated by making it less costly to write roachtests or to move more of those operations in-process thereby raising the bar for what should be a roachtest? Apologies if I'm being hand wavy but I couldn't find much on this topic.

Don't expect too much in 21.2 as we don't have any dedicated staffing to make anything happen. What we want to figure out is where the journey should go, but that doesn't buy you much in the long term. I definitely take the view though that we should write roachtests sparingly, resorting to them only when copious amounts of in-process testing are available. Then again, each major feature also needs to have the tires kicked over long time scales with real data, and cluster-to-cluster replication would definitely still want one. The rethinking of roachtests is primarily concerned with improving the ergonomics (lots of things are just not possible now or are very cumbersome) and reducing the cost of ownership of a roachtest.

**Comment by tbg** [comment]
For my education, could you point me at the in-process tests you have? I'm curious what TestCluster-to-TestCluster replication looks like.

**Comment by pbardea** [comment]
> For my education, could you point me at the in-process tests you have? I'm curious what TestCluster-to-TestCluster replication looks like.

Our E2E testing so far is pretty limited, but https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/streamingccl/streamingest/stream_ingestion_job_test.go#L39-L132 is running the flow from one test-cluster to another. Test here mostly are testing the happy path today, but we're working on adding more in-process tests here as well.

> I definitely take the view though that we should write roachtests sparingly, resorting to them only when copious amounts of in-process testing are available.

I also wanted to point to mixed-version testing, which are typically more short-lived tests which, based on my understanding, we need to resort back to roachtests (and when this feature is available in several versions is something we'll also want to test with roachtests).

Related to this conversation (but don't want to go too far on a tangent here and derail this thread), do we expect test-servers  to be a light-enough wrapper around actual servers that we can use them for tests around resilience against node failures? This is also something I've been reaching for roachtests for recently.

**Comment by tbg** [comment]
Thanks! Those look good.

> I also wanted to point to mixed-version testing, which are typically more short-lived tests which, based on my understanding, we need to resort back to roachtests (and when this feature is available in several versions is something we'll also want to test with roachtests).

Yes, mixed-version testing I fear we won't be able to replicate in-process anytime soon, so reaching for roachtests here is fine. I think the two things to do are a) add a "small" step to the existing acceptance/version-upgrade test, i.e. a quick sanity check, and b) a long-running mixed version test with, say, chaos, or something like that, which exclusively focuses on the feature at hand. So in a sense, you wouldn't write a new short-running mixed-version roachtest, rather just hook into the existing one.

**Comment by tbg** [comment]
> Related to this conversation (but don't want to go too far on a tangent here and derail this thread), do we expect test-servers to be a light-enough wrapper around actual servers that we can use them for tests around resilience against node failures? This is also something I've been reaching for roachtests for recently.

Yes, definitely! We recently added the capability to stop and restart nodes "properly" (it's still not 100% there, but close enough). There is also a HybridLogicalClock that you can use to introduce a clock jump to avoid having to sit out the lease expirations (i.e. to keep the test snappy).

File: pkg/cmd/roachprod/install/cluster_synced.go
Comment (irfansharif) [inline pkg/cmd/roachprod/install/cluster_synced.go:1742]: This comment could do with updating.

File: pkg/cmd/roachprod/install/cluster_synced.go
Comment (irfansharif) [inline pkg/cmd/roachprod/install/cluster_synced.go:1743]: s/targetNodeSpecified/allSpecifiedNodes?

File: pkg/cmd/roachprod/main.go
Comment (irfansharif) [inline pkg/cmd/roachprod/main.go:1056]: Could we keep our use of `verifyClusterName`, which gives us nice errors+suggestions if improper? Something like splitting args[0] on ":" should suffice, and passing the first argument to verify.

File: pkg/cmd/roachprod/main.go
Comment (irfansharif) [inline pkg/cmd/roachprod/main.go:126]: Name the return arguments. Also, what's "name"? how about clusterName for the cluster name return type? And parseNodesAndClusterName for this helper + s/name/arg for the input?

File: pkg/cmd/roachprod/install/cockroach.go
Comment (irfansharif) [inline pkg/cmd/roachprod/install/cockroach.go:476]: I mean, wouldn't we auto-init regardless of how the join flags are constructed, as long as --skip-init isn't specified? I'm not sure I follow.

**Review by irfansharif — APPROVED** [review]
LGTM, but heads up, roachprod changes tend to have a large blast radius that isn't necessarily caught in CI. I'll encourage you to keep an eye out for failures over the next few days and address it if they come up (if any, I'm guessing they'd be around doubly init-ing nodes or not initing them at all).

**Comment by tbg** [comment]
> The motivation for writing a roachtest was to be able to run the replication logic over a longer period of time with a workload running on the tenant to stress the streaming, ingestion, and cutover.

Having mulled this over a bit more, I still think the team would still get a lot more benefit from walking down this path using in-process testing before jumping to roachtests. I'm perfectly happy to entertain long-running in-process tests that have multiple clusters communicate and work through the implications of introducing them. I worry that introducing the concept of running multiple clusters in roachtests will only degrade the already shaky foundation it provides while piling up debt that we will eventually need to pay down. No such debt will be incurred using in-process testing.

I acknowledge that ultimately something like this needs to be tested at scale, but don't you expect to find tons of issues doing everything the roachtests would do in-process at smaller scale? It will be a lot easier to fix these issues there and then we're left testing only mixed versions and large scales in roachtest, and this buys us some time. I worry that if the team is leaning on this roachtest so early in the cycle, this will be where you find issues and this will be where you fix them.

Also, as an aside, introducing multiple clusters will not work well with roachtest's already limited observability. We pull a debug.zip from the first node we can reach, but with multiple clusters you need to pull multiple debug.zips. We thus need to bake the nodeID -> cluster mapping into the test harness or provide some way for roachtest to figure out this mapping by introspecting the vms.

**Comment by irfansharif** [comment]
Would a roachtest that makes use of multiple roachprod clusters be an acceptable stop-gap? It still breaks precedence, but at least it's contained at the level of the roachtest being written.

**Comment by tbg** [comment]
I think it's acceptable (and ultimately I don't want to wield veto power here) but I think the suggestion still stands regardless, that it may not be useful to go for a roachtest at this point. That's not for me to decide but it doesn't look that the in-process tests are doing very interesting things yet.

**Comment by gh-casper** [comment]
@tbg Hi Tobias. This is Casper from CDC and I'm working on tenant streaming project. I want to bring this up again as now we are trying to seriously roachtest the cluster to cluster streaming to bring this feature to GA in 22.2. I admit that I can still do some interesting testing in in-process tests, but it makes sense to have it ready to support multiple crdb clusters in single roachtest sooner than later.

I'm wondering if this PR is the recommended way to support two CRDB clusters in the same roachprod cluster. Have we considered making some changes in roachtest library level to support two roachprod clustes in single roachtest? After we agree on some approach, should test eng team take it over or CDC team can take it with test eng team's guidance? I'm happy with either way.

**Comment by tbg** [comment]
I'm fine with the rough approach this PR took. In terms of the final API in roachtest something like this makes sense:

```go
// two clusters: nodes [1,2,3], nodes [4,5,6]
c.Start(c.Range(1,3), option.DefaultStartOpts())
startOpts := option.DefaultStartOpts()
startOpts.RoachprodOpts.JoinTarget = 4 // <-- this is new
c.Start(c.Range(4,6), startOpts)
```

So we only need to add `JoinTarget` support to `roachprod`. The default `JoinTarget` is 1 (with the zero value being equivalent to 1). We apply the auto-init behavior to that node as well, i.e. if we start that node & it isn't initialized then we init it. The other nodes point at this node. So we're strictly generalizing the n1 behavior to other nodes when requested.

cc @srosenberg

**Comment by srosenberg** [comment]
If the immediate goal is to support multiple clusters without changing the roachtest framework, then the above API seems like the cleanest (and least invasive) option. I think we should also add `InitTarget` and make it explicit,

```
// two clusters: nodes [1,2,3], nodes [4,5,6]

// start the first cluster
startOpts.RoachprodOpts.JoinTarget = 1
startOpts.RoachprodOpts.InitTarget = 1

// start the second cluster
startOpts.RoachprodOpts.InitTarget = 4
startOpts.RoachprodOpts.JoinTarget = 4
c.Start(c.Range(4,6), startOpts)
```

Having said that, I see this option only as a stopgap. It's a rather leaky abstraction atop roachprod's single cluster API; e.g., now we have to use node index to execute commands against a specific cluster. Instead, we should add support for multiple `ClusterSpec`s in the roachtest framework. On a first approximation, it doesn't seem like a ton of changes; roughly, `test_runner` just needs to be extended to provision a cluster for each `ClusterSpec`. This way, we get multiple clusters which are not only physically disjoint but also logically disjoint; it also simplifies the test harness.

I am fine with the roachprod hack to unblock multi-tenant streaming if we agree that `JoinTarget` (nor `InitTarget`) will likely be deprecated when we refactor the roachtest to use multiple `ClusterSpec`s.

**Comment by gh-casper** [comment]
Sounds good! I'll do with JoinTarget and InitTarget solution first to get a quick win, then refactor the roachtest to use multiple ClusterSpecs at some point.

**Pull request closed without merging**
